### PR TITLE
APB-46,AIP006 and AIP007

### PR DESCRIPTION
**Pull request by AionJayT**
This PR updated the solidity compiler for supporting EDVerify, Blake2b hash and Transaction Hash.

### CODE REVIEW

File: solidity/libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -2191,8 +2191,11 @@ string FunctionType::identifier() const
 	case Kind::Selfdestruct: id += "selfdestruct"; break;
 	case Kind::Revert: id += "revert"; break;
 	case Kind::ECRecover: id += "ecrecover"; break;
+	case Kind::EDVerify: id += "edverify"; break;
 	case Kind::SHA256: id += "sha256"; break;
 	case Kind::RIPEMD160: id += "ripemd160"; break;
+        case Kind::BLAKE2b256: id += "black2b256"; break;
+        case Kind::TXHash: id += "transactionhash"; break;
```
Comment (aion-kelvin) [inline solidity/libsolidity/ast/Types.cpp]: inconsistent indentation

File: solidity/libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -2515,8 +2518,11 @@ bool FunctionType::isBareCall() const
 	case Kind::BareCallCode:
 	case Kind::BareDelegateCall:
 	case Kind::ECRecover:
+	case Kind::EDVerify:
 	case Kind::SHA256:
 	case Kind::RIPEMD160:
+        case Kind::BLAKE2b256:
```
Comment (aion-kelvin) [inline solidity/libsolidity/ast/Types.cpp]: inconsistent indentation

File: solidity/libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -2554,8 +2560,11 @@ bool FunctionType::isPure() const
 	return
 		m_kind == Kind::SHA3 ||
 		m_kind == Kind::ECRecover ||
+		m_kind == Kind::EDVerify ||
 		m_kind == Kind::SHA256 ||
 		m_kind == Kind::RIPEMD160 ||
+                m_kind == Kind::BLAKE2b256 ||
```
Comment (aion-kelvin) [inline solidity/libsolidity/ast/Types.cpp]: inconsistent indentation

File: solidity/libsolidity/codegen/ExpressionCompiler.cpp [hunk 4]
```diff
@@ -823,11 +823,19 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
...
 		{
 			_functionCall.expression().accept(*this);
-			static const map<FunctionType::Kind, u128> contractAddresses{{FunctionType::Kind::ECRecover, 1},
-															   {FunctionType::Kind::SHA256, 2},
-															   {FunctionType::Kind::RIPEMD160, 3}};
+			static const map<FunctionType::Kind, u128> contractAddresses{
+				{FunctionType::Kind::ECRecover, 1},
+				{FunctionType::Kind::SHA256, 2},
+				{FunctionType::Kind::RIPEMD160, 3},
+				{FunctionType::Kind::EDVerify, 16},
```
Comment (aion-kelvin) [inline solidity/libsolidity/codegen/ExpressionCompiler.cpp]: What do those numbers mean?  (16, 17, 18)

Comment (AionJayT) [reply solidity/libsolidity/codegen/ExpressionCompiler.cpp]: the contact address offset will be present in the Aion precompile contract

File: modFastVM/build.gradle [hunk 5]
```diff
@@ -9,6 +9,9 @@ ext.moduleName = 'aion.fastvm'
 
 test.dependsOn copyNativeLibsForModuleTests
 
+targetCompatibility = 10
+sourceCompatibility = 10
```
Comment (yulongaion) [inline modFastVM/build.gradle]: 11?

Comment (AionJayT) [reply modFastVM/build.gradle]: this is for the gradle testing issue, must set to the ver10 so far

File: solidity/libsolidity/ast/Types.cpp (hunk 1 above)
Comment (AionJayT) [reply solidity/libsolidity/ast/Types.cpp]: fixed

File: solidity/libsolidity/ast/Types.cpp (hunk 2 above)
Comment (AionJayT) [reply solidity/libsolidity/ast/Types.cpp]: fixed

File: solidity/libsolidity/ast/Types.cpp (hunk 3 above)
Comment (AionJayT) [reply solidity/libsolidity/ast/Types.cpp]: fixed

File: solidity/libsolidity/analysis/GlobalContext.cpp [hunk 6]
```diff
@@ -64,6 +68,8 @@ m_magicVariables(vector<shared_ptr<MagicVariableDeclaration const>>{make_shared<
 							make_shared<FunctionType>(strings(), strings{"bytes32"}, FunctionType::Kind::SHA256, true)),
 					make_shared<MagicVariableDeclaration>("ecrecover",
 							make_shared<FunctionType>(strings{"bytes32", "uint8", "bytes32", "bytes32"}, strings{"address"}, FunctionType::Kind::ECRecover)),
+                    make_shared<MagicVariableDeclaration>("edverify",
+                            make_shared<FunctionType>(strings{"bytes32", "bytes32", "bytes32", "bytes32"}, strings{"address"}, FunctionType::Kind::EDVerify)),
```
Comment (AlexandraRoatis) [inline solidity/libsolidity/analysis/GlobalContext.cpp]: formatting

**Pull request merged**